### ISSUE TITLE
Tool fails when NetworkManager is running

### ISSUE THREAD

**Issue by sophron** [wifiphisher/wifiphisher#639]
After #610 the tool throws an exception ('Interface "ALL" is controlled by NetworkManager.') whenever the NetworkManager is running, even when the user does not use the Internet sharing feature. 

When we don't want to share the Internet connection from a wireless interface, we should be fine with the NetworkManager managing our interfaces as long as they are not connected on the Internet.

@anakin1028 If this is a trivial fix, feel free to push directly to `master`.

**Comment by anakin1028** [comment]
@sophron do you mean when NetworkManager is on, we should run the original flow just choose those cards which are managed by NM and kill the interfering processes [i.e. NetworkManager] as usual?

**Comment by sophron** [comment]
Yes, I have to rephrase this.

If NetworkManager is enabled and our network interfaces are not connected on the Internet, do we have any issues? If the answer is no, we are fine leaving NetworkManager alive. 

I believe that's not the case though.  NetworkManager causes trouble (e.g. keeps connecting interfaces on the Internet due to AUTO-CONNECT flag), and we should kill it. We can kill it with the usual ways (airmon-ng, nmcli) or even better, use D-Bus to disable wireless.

For example,

```shell
$ dbus-send --system --print-reply \
            --dest=org.freedesktop.NetworkManager \
            /org/freedesktop/NetworkManager \
            org.freedesktop.DBus.Properties.Set \
            string:"org.freedesktop.NetworkManager" \
            string:"WirelessEnabled" \
            variant:boolean:false
```

And then bring the tool back to life when the tool shutdowns.

Again, this is the case when we DON'T use the Internet sharing through Wi-Fi feature.

**Comment by anakin1028** [comment]
@sophron thanks for the guidance : ) While it seems trivial to fix this, I still write some code to disable the `NetworkManager` and recover it when `network_manager.on_exit` and issue a PR for this fix. The function tests for with internet access and without internet access are all looks good for me.

**Comment by MrBradleyB** [comment]
Hey guys,
Finished looks awesome but I am getting that same error. I am new to Kali and I don't know how to turn off the network config. Any help appreciated, I'm running amd64 Kali live off of an sd card.

**Comment by anakin1028** [comment]
@MrBradleyB we'll merge this fix soon. 

Currently you can set the adapter card to unmanaged in file: `/etc/NetworkManager/NetworkManager.conf` . 

Add the `[keyfile]` section like the following [note: change the mac address to your adapter card]
```
[keyfile]
unmanaged-devices=mac:38:ec:11:00:00:00;mac:66:77:88:99:00:aa
```

**Comment by andronixist** [comment]
I believe I'm having same issue.  Hope you can help.  Thanks

root@myhostname:~# wifiphisher
[*] Starting Wifiphisher 1.3GIT at 2017-06-14 14:13
[!] Interface "ALL" is controlled by NetworkManager.You need to manually set the devices that should be ignored by NetworkManager using the keyfile plugin (unmanaged-directive). For example, '[keyfile] unmanaged-devices=interface-name:"ALL"' needs to be added in your NetworkManager configuration file.
[+] Captured credentials:
[!] Closing
root@myhostname:~#

**Comment by PunkHack** [comment]
Hi @anakin1028 i tried to do what you say: Add the [keyfile] section like the following [note: change the mac address to your adapter card]
But it doesn't work. 
This is how my NetworkManager.conf looks:
------------------------------------------------------------------
[main]
plugins=ifupdown,keyfile

[keyfile]
unmanaged-devices=mac:6c:62:6d:2c:34:4b

[ifupdown]
managed=false
------------------------------------------------------------------

Im also try to disable the networkmanager by dbus but when i type "setup network_manager object" says "bash: setup: no se encontró la orden" ("Bash: Configuration: Order not found").

What am i doing wrong?

**Comment by anakin1028** [comment]
@PunkHack can you type the command `nmcli dev` and post the log?

**Comment by PunkHack** [comment]
DISPOSITIVO  TIPO      ESTADO       CONEXIÓN            
eth0         ethernet  conectado    Conexión cableada 1 
wlan0        wifi      conectado    Rock Family         
lo           loopback  sin gestión  --

**Comment by anakin1028** [comment]
is wlan0's mac address equal to `6c:62:6d:2c:34:4b`?

**Comment by PunkHack** [comment]
I think so! I mean, when i tipe ifconfig this pop up

ether 6c:62:6d:2c:34:4b  txqueuelen 1000  (Ethernet)

**Comment by anakin1028** [comment]
can you post the log for `iw dev` ?

**Comment by PunkHack** [comment]
I guess this 48:5d:60:e8:33:6f is mi real mac. 

I put it on the NetworkManager and the problem disappear, but now says this when i try to open the wifiphisher:
Failed to find an interface with AP mode

**Comment by anakin1028** [comment]
@PunkHack it is because you have only one card. try to use this command:
`wifiphisher -nJ -p oauth-login`

**Comment by sophron** [comment]
@PunkHack This is because none of your wireless cards support AP mode. This is necessary for spawning the rogue Access Point.

@anakin1028 Can you add the check for the required h/w before the check of NetworkManager? Someone that wants to use the Internet sharing feature should get a warning that there's hardware missing the earliest possible.

**Comment by PunkHack** [comment]
@anakin1028 
@sophron 
No. I just had to tipe wifiphisher --nojammin instead wifiphisher and thats it, i finally made it!!! 

You're the best guys.. I'd been searching an answer for a week and thanks to you i could run the wifiphisher!

**Comment by sophron** [comment]
@PunkHack Good to hear. Two things.

1) Can you do a `iw dev`? This should show your wireless devices. Then, can you paste the output of `iw dev phy# info` here (where `phy#` the devices shown earlier)?

2) Can you run `git show | grep commit` command and paste here the output?

**Comment by anakin1028** [comment]
@sophron  can you rephrase what kind of h/w check we should do? Do you mean if the remaining adapters set (i.e. excluding the internet access one) contain the AP or Deauth features ?

**Comment by PunkHack** [comment]
@anakin1028 
@sophron 
No. I just had to tipe wifiphisher --nojammin instead wifiphisher and thats it, i finally made it!!! 

You're the best guys.. I'd been searching an answer for a week and thanks to you i could run the wifiphisher!

**Comment by sophron** [comment]
The tool should first check if the network adapters are compatible (at least one supports monitor mode and at least one supports AP mode). 

After that, if Internet sharing through wireless is enabled, the tool should check if there is a third (separate) interface that is also not managed by NetworkManager.

A bug that I understood from @PunkHack comments, is that the "NetworkManager" check happens before the h/w check.

**Comment by PunkHack** [comment]
root@kali:~# iw dev
phy#0
	Interface wlan0
		ifindex 3
		wdev 0x1
		addr 48:5d:60:e8:33:6f
		type managed
		txpower 0.00 dBm
root@kali:~# iw dev phy#0 info
command failed: No such device (-19)
root@kali:~# git show
fatal: Not a git repository (or any of the parent directories): .git
root@kali:~# grep commit


With the grep commit the terminal isnt doing anything.
Also, cant turn on mi wifi, why is that? I mean, the wifi is on but doesn't find any connection. I guess it because the power is 0, but why?

**Comment by anakin1028** [comment]
@PunkHack you should type `iw dev phy0 info`

**Comment by PunkHack** [comment]
root@kali:~# iw dev
phy#0
	Interface wlan0
		ifindex 3
		wdev 0x1
		addr 48:5d:60:e8:33:6f
		type managed
		txpower 0.00 dBm
root@kali:~# iw dev phy0 info
command failed: No such device (-19)
root@kali:~#

**Comment by anakin1028** [comment]
sorry I have fixed the typo before  -- > should be `iw phy phy0 info`

**Comment by PunkHack** [comment]
root@kali:~# iw phy phy0 info
Wiphy phy0
	max # scan SSIDs: 4
	max scan IEs length: 2257 bytes
	max # sched scan SSIDs: 0
	max # match sets: 0
	max # scan plans: 1
	max scan plan interval: -1
	max scan plan iterations: 0
	Retry short limit: 7
	Retry long limit: 4
	Coverage class: 0 (up to 0m)
	Device supports RSN-IBSS.
	Device supports AP-side u-APSD.
	Device supports T-DLS.
	Supported Ciphers:
		* WEP40 (00-0f-ac:1)
		* WEP104 (00-0f-ac:5)
		* TKIP (00-0f-ac:2)
		* CCMP-128 (00-0f-ac:4)
		* CCMP-256 (00-0f-ac:10)
		* GCMP-128 (00-0f-ac:8)
		* GCMP-256 (00-0f-ac:9)
		* CMAC (00-0f-ac:6)
		* CMAC-256 (00-0f-ac:13)
		* GMAC-128 (00-0f-ac:11)
		* GMAC-256 (00-0f-ac:12)
	Available Antennas: TX 0x1 RX 0x3
	Configured Antennas: TX 0x1 RX 0x3
	Supported interface modes:
		 * IBSS
		 * managed
		 * AP
		 * AP/VLAN
		 * WDS
		 * monitor
		 * mesh point
		 * P2P-client
		 * P2P-GO
		 * outside context of a BSS
	Band 1:
		Capabilities: 0x11ce
			HT20/HT40
			SM Power Save disabled
			RX HT40 SGI
			TX STBC
			RX STBC 1-stream
			Max AMSDU length: 3839 bytes
			DSSS/CCK HT40
		Maximum RX AMPDU length 65535 bytes (exponent: 0x003)
		Minimum RX AMPDU time spacing: 8 usec (0x06)
		HT TX/RX MCS rate indexes supported: 0-7
		Bitrates (non-HT):
			* 1.0 Mbps
			* 2.0 Mbps (short preamble supported)
			* 5.5 Mbps (short preamble supported)
			* 11.0 Mbps (short preamble supported)
			* 6.0 Mbps
			* 9.0 Mbps
			* 12.0 Mbps
			* 18.0 Mbps
			* 24.0 Mbps
			* 36.0 Mbps
			* 48.0 Mbps
			* 54.0 Mbps
		Frequencies:
			* 2412 MHz [1] (14.0 dBm)
			* 2417 MHz [2] (15.0 dBm)
			* 2422 MHz [3] (15.0 dBm)
			* 2427 MHz [4] (15.0 dBm)
			* 2432 MHz [5] (15.0 dBm)
			* 2437 MHz [6] (15.0 dBm)
			* 2442 MHz [7] (15.0 dBm)
			* 2447 MHz [8] (15.0 dBm)
			* 2452 MHz [9] (15.0 dBm)
			* 2457 MHz [10] (15.0 dBm)
			* 2462 MHz [11] (15.0 dBm)
			* 2467 MHz [12] (15.0 dBm) (no IR)
			* 2472 MHz [13] (14.0 dBm) (no IR)
			* 2484 MHz [14] (17.0 dBm) (no IR)
	Supported commands:
		 * new_interface
		 * set_interface
		 * new_key
		 * start_ap
		 * new_station
		 * new_mpath
		 * set_mesh_config
		 * set_bss
		 * authenticate
		 * associate
		 * deauthenticate
		 * disassociate
		 * join_ibss
		 * join_mesh
		 * remain_on_channel
		 * set_tx_bitrate_mask
		 * frame
		 * frame_wait_cancel
		 * set_wiphy_netns
		 * set_channel
		 * set_wds_peer
		 * tdls_mgmt
		 * tdls_oper
		 * probe_client
		 * set_noack_map
		 * register_beacons
		 * start_p2p_device
		 * set_mcast_rate
		 * channel_switch
		 * set_qos_map
		 * connect
		 * disconnect
	Supported TX frame types:
		 * IBSS: 0x00 0x10 0x20 0x30 0x40 0x50 0x60 0x70 0x80 0x90 0xa0 0xb0 0xc0 0xd0 0xe0 0xf0
		 * managed: 0x00 0x10 0x20 0x30 0x40 0x50 0x60 0x70 0x80 0x90 0xa0 0xb0 0xc0 0xd0 0xe0 0xf0
		 * AP: 0x00 0x10 0x20 0x30 0x40 0x50 0x60 0x70 0x80 0x90 0xa0 0xb0 0xc0 0xd0 0xe0 0xf0
		 * AP/VLAN: 0x00 0x10 0x20 0x30 0x40 0x50 0x60 0x70 0x80 0x90 0xa0 0xb0 0xc0 0xd0 0xe0 0xf0
		 * mesh point: 0x00 0x10 0x20 0x30 0x40 0x50 0x60 0x70 0x80 0x90 0xa0 0xb0 0xc0 0xd0 0xe0 0xf0
		 * P2P-client: 0x00 0x10 0x20 0x30 0x40 0x50 0x60 0x70 0x80 0x90 0xa0 0xb0 0xc0 0xd0 0xe0 0xf0
		 * P2P-GO: 0x00 0x10 0x20 0x30 0x40 0x50 0x60 0x70 0x80 0x90 0xa0 0xb0 0xc0 0xd0 0xe0 0xf0
		 * P2P-device: 0x00 0x10 0x20 0x30 0x40 0x50 0x60 0x70 0x80 0x90 0xa0 0xb0 0xc0 0xd0 0xe0 0xf0
	Supported RX frame types:
		 * IBSS: 0x40 0xb0 0xc0 0xd0
		 * managed: 0x40 0xd0
		 * AP: 0x00 0x20 0x40 0xa0 0xb0 0xc0 0xd0
		 * AP/VLAN: 0x00 0x20 0x40 0xa0 0xb0 0xc0 0xd0
		 * mesh point: 0xb0 0xc0 0xd0
		 * P2P-client: 0x40 0xd0
		 * P2P-GO: 0x00 0x20 0x40 0xa0 0xb0 0xc0 0xd0
		 * P2P-device: 0x40 0xd0
	software interface modes (can always be added):
		 * AP/VLAN
		 * monitor
	valid interface combinations:
		 * #{ managed } <= 2048, #{ AP, mesh point } <= 8, #{ P2P-client, P2P-GO } <= 1,
		   total <= 2048, #channels <= 1, STA/AP BI must match
		 * #{ WDS } <= 2048,
		   total <= 2048, #channels <= 1, STA/AP BI must match
	HT Capability overrides:
		 * MCS: ff ff ff ff ff ff ff ff ff ff
		 * maximum A-MSDU length
		 * supported channel width
		 * short GI for 40 MHz
		 * max A-MPDU length exponent
		 * min MPDU start spacing
	Device supports TX status socket option.
	Device supports HT-IBSS.
	Device supports SAE with AUTHENTICATE command
	Device supports low priority scan.
	Device supports scan flush.
	Device supports AP scan.
	Device supports per-vif TX power setting
	P2P GO supports CT window setting
	Driver supports full state transitions for AP/GO clients
	Driver supports a userspace MPM
	Device supports active monitor (which will ACK incoming frames)
	Driver/device bandwidth changes during BSS lifetime (AP/GO mode)
	Device supports configuring vdev MAC-addr on create.
root@kali:~#

**Comment by anakin1028** [comment]
@sophron I still don't understand.

Our card selection considers two factors : 
One is `supported mode` (i.e. like you mentioned `h/w mode`) and the other is whether the card is `managed by NetworkManager`. We don't check the NetworkManager before checking h/w supported mode but check that two factors in the meantime. If the card support the right mode, but is managed by NetworkManager we'll raise the exception for it. Do you think any flaws for this algorithm?

The problem for this issue is that during the card selection the `NetworkManager` is on, so all cards are being managed by NetworkManager and they will all fail in card selection algorithm. For the newest PR we merge, we'll solve this problem since we disable the NetworkManager; however if users want to use wireless connection, they should take care about the `managed` property of the cards.

**Comment by anakin1028** [comment]
@PunkHack yes, good~ your card looks good since it both support monitor and AP mode
```
Supported interface modes:
* IBSS
* managed
* AP
* AP/VLAN
* WDS
* monitor
* mesh point
```

**Comment by PunkHack** [comment]
@anakin1028 Exactly but i still cant find the way to turn the txpower on again. Still in 0 and obviously cant find any network!

**Comment by anakin1028** [comment]
I know what's the problem. You mean now you want to use your card to connect to WiFi? If so you have to remove the keyfile section you have added in your NetworkManager.conf file since this section will cause NetworkManager not control your adapter.

**Comment by anakin1028** [comment]
And can you pull the latest commit from `git` ? we have fixed this issue now, you don't have to modify the `keyfile` section if you don't want to use internet access when you use `wifiphisher`

**Comment by PunkHack** [comment]
Ok let me see if i understand. When im going to use wifi normally (to check some facebook or anything) the networkmanager can't content the keyfile, right? Then when i wanna use the wifiphisher i have to add the keyfile again?

This is the git you ask me: root@kali:~# git
usage: git [--version] [--help] [-C <path>] [-c name=value]
           [--exec-path[=<path>]] [--html-path] [--man-path] [--info-path]
           [-p | --paginate | --no-pager] [--no-replace-objects] [--bare]
           [--git-dir=<path>] [--work-tree=<path>] [--namespace=<name>]
           <command> [<args>]

These are common Git commands used in various situations:

start a working area (see also: git help tutorial)
   clone      Clone a repository into a new directory
   init       Create an empty Git repository or reinitialize an existing one

work on the current change (see also: git help everyday)
   add        Add file contents to the index
   mv         Move or rename a file, a directory, or a symlink
   reset      Reset current HEAD to the specified state
   rm         Remove files from the working tree and from the index

examine the history and state (see also: git help revisions)
   bisect     Use binary search to find the commit that introduced a bug
   grep       Print lines matching a pattern
   log        Show commit logs
   show       Show various types of objects
   status     Show the working tree status

grow, mark and tweak your common history
   branch     List, create, or delete branches
   checkout   Switch branches or restore working tree files
   commit     Record changes to the repository
   diff       Show changes between commits, commit and working tree, etc
   merge      Join two or more development histories together
   rebase     Reapply commits on top of another base tip
   tag        Create, list, delete or verify a tag object signed with GPG

collaborate (see also: git help workflows)
   fetch      Download objects and refs from another repository
   pull       Fetch from and integrate with another repository or a local branch
   push       Update remote refs along with associated objects

'git help -a' and 'git help -g' list available subcommands and some
concept guides. See 'git help <command>' or 'git help <concept>'
to read about a specific subcommand or concept.
root@kali:~#

**Comment by anakin1028** [comment]
@PunkHack yes right for the first part. You don't have to use keyfile when you want this card to connect to WiFi.
For the second part : you should go to your `wifiphisher` directory (i.e. The directory where you have cloned from github), and do what @sophron have said type the exactly as the follows (i.e. this is a single command don't separate as two commands like you have done before):
`git show | grep commit`

**Comment by PunkHack** [comment]
Di you mean this: 
root@kali:~/github/wifiphisher# git show | grep commit
commit b7650289b765c01a6ab22de1bca0a354ce4265d0
root@kali:~/github/wifiphisher#

**Comment by anakin1028** [comment]
yes you are right. you can then type the command `git pull origin master` to get the latest commit and this time don't type `keyfile` section and see if `wifiphisher` works as you have typed `keyfile` section before.

**Comment by PunkHack** [comment]
root@kali:~/github/wifiphisher# git pull origin master
remote: Counting objects: 32, done.
remote: Compressing objects: 100% (3/3), done.
remote: Total 32 (delta 20), reused 22 (delta 20), pack-reused 9
Unpacking objects: 100% (32/32), done.
From https://github.com/wifiphisher/wifiphisher
 * branch            master     -> FETCH_HEAD
   b765028..db50968  master     -> origin/master
Updating b765028..db50968
error: Your local changes to the following files would be overwritten by merge:
	setup.py
Please commit your changes or stash them before you merge.
Aborting

And when i try to open wifiphisher say the same error: Interface "ALL" is controlled by NetworkManager.

**Comment by anakin1028** [comment]
your `git pull` command has been failed since you have modified your files in your folder. You should backup your changes and revert it to the original state and then run the `git pull origin master` again. 

Maybe have a study about `Git` will be useful to you too.

**Comment by PunkHack** [comment]
Maybe its the sleep talking but i have no idea what are you talking about =/

How can i backup and revert the git pull?

By the way, as you can see I'm totally new at this. I really appreciate your patience and your help!

**Comment by anakin1028** [comment]
You can use `git stash save this my temp save` to store your temporary change. The run:
`git fetch` and then
`git merge origin master`

**Comment by PunkHack** [comment]
@anakin1028 Alright man I'm done. I have to wake up in 3 hours but i'll be back tomorrow to see if i can make it.

Once again, thanks for your time and your help... Same to you @sophron, thanks a lot!

**Comment by PunkHack** [comment]
@anakin1028 Hi man, i'm back.

This is what says when i type 
root@kali:~/github# git fetch
fatal: Not a git repository (or any of the parent directories): .git

**Comment by blackHatMonkey** [comment]
@PunkHack If you don't care about the changes you made to the `setup.py` file then you can simply run the following command:

**Note: make sure you are in the `wifiphisher` directory** 

```bash
git fetch && git reset --hard origin master
```

**Comment by PunkHack** [comment]
Done!! Thanks @blackHatMonkey, as you say, i wasn't in the right folder.

I didn't follow your command, i tried first with this:
"You can use git stash save this my temp save to store your temporary change. The run:
git fetch and then
git merge origin master". 

Just like @anakin1028 says and works perfectly, now i can use wifi and wifiphisher in the same time.

**Comment by anakin1028** [comment]
@PunkHack  congrats : )

**Comment by PunkHack** [comment]
Guys one last thing if you can help me. Did you know some wifiphisher tutorial or something that can be helpful for me?

Its because my wifiphisher works fine now but i'd been doing some research to learn how to crack wifi passwords but when i tried to do "network manager connect or firmware pluggin update" doesn't deactivates any device on my network. So I'm basically stock in there .
Guys one last thing if you can help me. Did you know some wifiphisher tutorial or something that can be helpful for me?

Its because my wifiphisher works fine now but i'd been doing some research to learn how to crack wifi passwords but when i tried to do "network manager connect or firmware pluggin update" doesn't deactivates any device on my network. So I'm basically stuck in there .

**Comment by anakin1028** [comment]
@PunkHack you can read the `README` first. As far as I know, it seems that you only have one adatper card, so our tool will not lunch deauth attack to deauth the clients that connected to the targeted AP.

To lunch the deauth attack you should use the command : `wifiphisher -aI wlan0 -jI wlan1 -p firmware-upgrade` (i.e. wlan1 must have `monitor` mode and you can use the command `iw phy phy# info` to check that as we have showed you yesterday)

And the other thing you need to understand is that unless your target AP (i.e. the AP you selected to attack) is OPEN network, the victim will auto-connected to the fake-ap. So if the target-AP has security, you should manually select on your client devices (i.e. victims) to the fake-ap you have lunched. you can read #560 to give you a started point.
